pkg/stream/base_consumer.go

```diff
@@ -0,0 +1,363 @@
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-ticker.C:
+			c.processBatch(ctx)
+		}
+	}
+}
+
+func (c *baseConsumer) processBatch(ctx context.Context) {
+	c.m.Lock()
+	batch := c.batch
+	c.batch = make([]*Message, 0, c.settings.BatchSize)
+	c.m.Unlock()
+
+	c.doConsuming(ctx, batch)
+}
+
+func (c *baseConsumer) runCallback(ctx context.Context) error {
+	logger := c.logger.WithContext(ctx)
+	defer logger.Debug("runCallback is ending")
+
+	if runnable, ok := c.callback.(RunnableConsumerCallback); ok {
+		return runnable.Run(ctx)
+	}
+
+	return nil
+}
+
+func (c *baseConsumer) stopConsuming() error {
+	defer c.logger.Debug("stopConsuming is ending")
+
+	c.wg.Wait()
+	c.input.Stop()
+	c.cancel()
+
+	return nil
+}
+
+func (c *baseConsumer) runConsuming(ctx context.Context) error {
+	logger := c.logger.WithContext(ctx)
+	defer logger.Debug("runConsuming is ending")
+	defer c.wg.Done()
+
+	var ok bool
+	var msg *Message
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("return from consuming as the coffin is dying")
+
+		case msg, ok = <-c.input.Data():
+		}
+
+		if !ok {
+			c.processBatch(ctx)
+			return nil
+		}
+
+		c.m.Lock()
+		c.batch = append(c.batch, msg)
+		c.m.Unlock()
+
+		if len(c.batch) < c.settings.BatchSize {
+			continue
+		}
+
+		c.processBatch(ctx)
+	}
+}
+
+func (c *baseConsumer) doConsuming(ctx context.Context, batch []*Message) {
+	defer c.recover()
+
+	if len(batch) == 0 {
+		return
+	}
+
+	logger := c.logger.WithContext(ctx)
+
+	var err error
+	var span tracing.Span
+	models := make([]interface{}, len(batch), len(batch))
+	attributes := make([]map[string]interface{}, len(batch), len(batch))
+	spans := make([]tracing.Span, len(batch), len(batch))
+
+	for i, msg := range batch {
+		models[i] = c.callback.(BaseConsumerCallback).GetModel(msg.Attributes)
+
+		ctx, attributes[i], err = c.encoder.Decode(ctx, msg, models[i])
+		if err != nil {
+			logger.WithContext(ctx).Error(err, "an error occurred during the batch consume operation")
+			continue
+		}
+
+		ctx, span = c.tracer.StartSpanFromContext(ctx, c.id)
+		spans[i] = span
+	}
+
+	defer func() {
+		for _, span := range spans {
+			span := span
+			span.Finish()
+		}
+	}()
+
+	processed := 0
+	switch c.callback.(type) {
+	case BatchConsumerCallback:
+		processed = c.callBatchCallback(ctx, batch, models, attributes)
+	case ConsumerCallback:
+		for i := range models {
+			processed += c.callSingleCallback(ctx, batch[i], models[i], attributes[i])
+		}
+	}
+
+	atomic.AddInt32(&c.processed, int32(processed))
+	c.mw.WriteOne(&mon.MetricDatum{
+		MetricName: metricNameConsumerProcessedCount,
+		Value:      float64(processed),
+	})
+}
+
+func (c *baseConsumer) recover() {
+	err := coffin.ResolveRecovery(recover())
+	if err == nil {
+		return
+	}
+
+	c.logger.Error(err, err.Error())
+}
+
+func (c *baseConsumer) callBatchCallback(ctx context.Context, batch []*Message, models []interface{}, attributes []map[string]interface{}) int {
```

Comment:
The base consumer must not know about the consumption process, i.e. whether it's batch or not.
Those are implementation details of the actual consumers.
Please move anything related to that to the specific implementation.

Consider adding a new consumer type which does not fit the batch or single processing pattern, then you would need to touch the base consumer again.

Reply:
I think this requires moving of `doConsuming` aswell

Reply:
This is what we had before you told me to migrate the two. Maybe you want to finish the implementation yourself?